cpp/src/arrow/compute/exec/source_node.cc

```diff
@@ -107,19 +109,19 @@ struct SourceNode : ExecNode {
                 ExecBatch batch = std::move(*maybe_batch);
 
                 if (executor) {
-                  auto status =
-                      task_group_.AddTask([this, executor, batch]() -> Result<Future<>> {
-                        return executor->Submit([=]() {
-                          outputs_[0]->InputReceived(this, std::move(batch));
-                          return Status::OK();
-                        });
-                      });
+                  auto status = task_group_.AddTask([this, executor,
+                                                     batch]() -> Result<Future<>> {
+                    return executor->Submit([=]() {
+                      outputs_[0]->InputReceived(this, IdentityTask(std::move(batch)));
+                      return Status::OK();
+                    });
+                  });
                   if (!status.ok()) {
                     outputs_[0]->ErrorReceived(this, std::move(status));
                     return Break(total_batches);
                   }
                 } else {
-                  outputs_[0]->InputReceived(this, std::move(batch));
+                  outputs_[0]->InputReceived(this, IdentityTask(std::move(batch)));
```

Comment:
This would actually **not** create a task but forward to downstream like filter/project.